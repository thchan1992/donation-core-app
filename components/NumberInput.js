import { StyleSheet, Text, TextInput, View } from "react-native";

const NumberInput = ({
  value,
  setter,
  placeholder,
  height,
  maxWidth,
  minWidth,
  maxLength,
  isPound,
}) => {
  return (
    <View style={{ flexDirection: "row" }}>
      {isPound != null && <Text style={styles.poundsSign}>Â£</Text>}
      <TextInput
        textAlign={"center"}
        placeholderTextColor={"grey"}
        onChangeText={setter}
        maxLength={maxLength}
        style={[
          styles.input,
          {
            minheight: height,
            fontSize: height,
            maxWidth: maxWidth,
            minWidth: minWidth,
          },
        ]}
        value={value}
        placeholder={placeholder}
        keyboardType="numeric"
      />
    </View>
  );
};

export default NumberInput;

const styles = StyleSheet.create({
  input: {
    margin: 12,
    borderWidth: 1,
    padding: 10,
    borderColor: "#FF304F",
    borderRadius: 8,
    minheight: 50,
    fontSize: 50,
    color: "white",
    fontSize: 50,
    maxWidth: 200,
  },
  poundsSign: {
    fontSize: 50,
    color: "white",
    paddingTop: 20,
  },
});
