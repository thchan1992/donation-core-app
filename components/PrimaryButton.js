import { TouchableOpacity, Text, StyleSheet } from "react-native";
// PrimaryButton is a functional component that renders a button with a text and onPress functionality
const PrimaryButton = ({ text, onPress }) => {
  return (
    // TouchableOpacity is used to make the button responsive to touch/press events
    <TouchableOpacity style={styles.button} onPress={onPress}>
      <Text style={styles.text}>{text}</Text>
    </TouchableOpacity>
  );
};

export default PrimaryButton;

const styles = StyleSheet.create({
  button: {
    width: 150,
    height: 70,
    margin: 10,
    borderRadius: 8,
    backgroundColor: "#28C7FA",
    borderRadius: 8,
    borderWidth: 1.5,
    borderColor: "#FF304F",
  },
  text: {
    alignSelf: "center",
    padding: 16,
    fontWeight: "bold",
    fontSize: 30,
  },
});
