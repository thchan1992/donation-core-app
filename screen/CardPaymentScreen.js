import { View, Text, StyleSheet, Alert } from "react-native";
import { useState } from "react";
import NumberInput from "../components/NumberInput";
import PrimaryButton from "../components/PrimaryButton";

const CardPaymentScreen = ({ navigation, route }) => {
  const { charity, donateAmount } = route.params;
  const [cardNum, setCardNum] = useState(null);
  const [cvv, setCvv] = useState(null);
  const [expDate, setExpDate] = useState("");
  const fixCardNum = "1234889012345555";
  const fixCvv = "539";
  const fixExpDate = "02/23";

  const showAlert = () => {
    Alert.alert(
      "Warning",
      "The system does not recognise this credit card, please try again.",
      [{ text: "OK", onPress: () => console.log("OK Pressed") }]
    );
  };

  const handlePayment = () => {
    if (cardNum == fixCardNum && cvv == fixCvv && expDate == fixExpDate) {
      navigation.navigate("ReceiptScreen", {
        charity: charity,
        donateAmount: donateAmount,
      });
    } else {
      showAlert();
    }
  };

  const handleExpDate = (e) => {
    setExpDate(e);
    if (e.length == 2 && expDate.length == 1) {
      setExpDate((e) => e + "/");
    }
  };
  return (
    <View style={styles.container}>
      <Text style={styles.instrucText}>
        Your donation amount to {charity.name} is: {donateAmount}
      </Text>
      <NumberInput
        value={cardNum}
        setter={setCardNum}
        placeholder={"Card Number"}
        height={29}
        maxWidth={300}
        minWidth={300}
        maxLength={16}
      />
      <NumberInput
        value={cvv}
        setter={setCvv}
        placeholder={"CVV"}
        height={29}
        maxWidth={80}
        minWidth={80}
        maxLength={3}
      />
      <NumberInput
        value={expDate}
        setter={handleExpDate}
        placeholder={"EXP DATE"}
        height={29}
        maxWidth={150}
        minWidth={150}
        maxLength={5}
      />
      <PrimaryButton text="Confirm" onPress={handlePayment} />
    </View>
  );
};

const styles = StyleSheet.create({
  container: { flex: 1, justifyContent: "center", alignItems: "center" },
  instrucText: {
    alignSelf: "center",
    padding: 5,
    fontWeight: "bold",
    fontSize: 25,
    color: "white",
  },
});

export default CardPaymentScreen;
