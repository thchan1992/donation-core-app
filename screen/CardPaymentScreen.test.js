import renderer from "react-test-renderer";
import React from "react";
import CardPaymentScreen from "./CardPaymentScreen";
import NumberInput from "../components/NumberInput";
import PrimaryButton from "../components/PrimaryButton";
import { Alert } from "react-native";

describe("CardPaymentScreen", () => {
  let navigation;
  let route;

  beforeEach(() => {
    navigation = {
      navigate: jest.fn(),
    };
    route = {
      params: {
        charity: {
          name: "Test Charity",
          imageUrl: require("../assets/adaptive-icon.png"),
        },
        donateAmount: 50,
      },
    };
    jest.spyOn(Alert, "alert").mockImplementation(() => {});
  });

  it("calls navigate when payment is successful", () => {
    const component = renderer.create(
      <CardPaymentScreen navigation={navigation} route={route} />
    );
    const instance = component.root;
    const cardNumInput = instance.find(
      (node) =>
        node.type === NumberInput && node.props.placeholder === "Card Number"
    );
    cardNumInput.props.setter("1111111111111111");
    const cvvInput = instance.find(
      (node) => node.type === NumberInput && node.props.placeholder === "CVV"
    );
    cvvInput.props.setter("111");
    const expDateInput = instance.find(
      (node) =>
        node.type === NumberInput && node.props.placeholder === "EXP DATE"
    );
    expDateInput.props.setter("11/11");
    const confirmButton = instance.find(
      (node) => node.type === PrimaryButton && node.props.text === "Confirm"
    );
    confirmButton.props.onPress();
    expect(navigation.navigate).toHaveBeenCalledWith(
      "ReceiptScreen",
      route.params
    );
  });

  it("shows an alert when payment is not successful", () => {
    const component = renderer.create(
      <CardPaymentScreen navigation={navigation} route={route} />
    );
    const instance = component.root;
    const cardNumInput = instance.find(
      (node) =>
        node.type === NumberInput && node.props.placeholder === "Card Number"
    );
    cardNumInput.props.setter("1111111111111");
    const cvvInput = instance.find(
      (node) => node.type === NumberInput && node.props.placeholder === "CVV"
    );
    cvvInput.props.setter("111");

    const expDateInput = instance.find(
      (node) =>
        node.type === NumberInput && node.props.placeholder === "EXP DATE"
    );

    expDateInput.props.setter("11/11");
    const confirmButton = instance.find(
      (node) => node.type === PrimaryButton && node.props.text === "Confirm"
    );
    confirmButton.props.onPress();
    // check if alert is shown
    expect(Alert.alert).toHaveBeenCalled();
  });
});
