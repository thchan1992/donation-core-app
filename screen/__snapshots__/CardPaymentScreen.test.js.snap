// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardPaymentScreen renders correctly 1`] = `
<RCTScrollView
  collapsable={false}
  handlerTag={1}
  handlerType="NativeViewGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "white",
            "fontSize": 25,
            "fontWeight": "bold",
            "padding": 5,
          }
        }
      >
        Your donation amount to Test Charity is: Â£10. Please enter your credit card detail below.
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "height": 550,
          }
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            keyboardType="numeric"
            maxLength={16}
            onChangeText={[Function]}
            placeholder="Card Number"
            placeholderTextColor="grey"
            style={
              Array [
                Object {
                  "borderColor": "#FF304F",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "color": "white",
                  "fontSize": 50,
                  "margin": 12,
                  "maxWidth": 200,
                  "minheight": 50,
                  "padding": 10,
                },
                Object {
                  "fontSize": 29,
                  "maxWidth": 300,
                  "minWidth": 300,
                  "minheight": 29,
                },
              ]
            }
            textAlign="center"
            value={null}
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            keyboardType="numeric"
            maxLength={3}
            onChangeText={[Function]}
            placeholder="CVV"
            placeholderTextColor="grey"
            style={
              Array [
                Object {
                  "borderColor": "#FF304F",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "color": "white",
                  "fontSize": 50,
                  "margin": 12,
                  "maxWidth": 200,
                  "minheight": 50,
                  "padding": 10,
                },
                Object {
                  "fontSize": 29,
                  "maxWidth": 80,
                  "minWidth": 80,
                  "minheight": 29,
                },
              ]
            }
            textAlign="center"
            value={null}
          />
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            keyboardType="numeric"
            maxLength={5}
            onChangeText={[Function]}
            placeholder="EXP DATE"
            placeholderTextColor="grey"
            style={
              Array [
                Object {
                  "borderColor": "#FF304F",
                  "borderRadius": 8,
                  "borderWidth": 1,
                  "color": "white",
                  "fontSize": 50,
                  "margin": 12,
                  "maxWidth": 200,
                  "minheight": 50,
                  "padding": 10,
                },
                Object {
                  "fontSize": 29,
                  "maxWidth": 150,
                  "minWidth": 150,
                  "minheight": 29,
                },
              ]
            }
            textAlign="center"
            value=""
          />
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#28C7FA",
              "borderColor": "#FF304F",
              "borderRadius": 8,
              "borderWidth": 1.5,
              "height": 70,
              "margin": 10,
              "opacity": 1,
              "width": 150,
            }
          }
        >
          <Text
            style={
              Object {
                "alignSelf": "center",
                "fontSize": 30,
                "fontWeight": "bold",
                "padding": 16,
              }
            }
          >
            Confirm
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
